### PR TITLE
String input sanity when querying chargen template filename

### PR DESCRIPTION
**Pull request by Muffindrake**
Currently, the character generation allows you to write arbitrary files to the system using a string like

    ../../../../../../home/muffindrake/file

as a filename when prompted to do so during the template save dialogue.

While this is not really a major problem per se, the written template file will no longer be accessible to the game when written out of the templates directory in this manner, which I doubt was the intention of that code. It may also cause a major annoyance for operators of servers running this game that could find character template files in random location in a user's home. Furthermore, in the unlikely occasion that an important file is named with an extension `.template`, that file will be overwritten.

As such, I ask for sanity in that specific dialogue.

My commit checks if the string contains any forward slash and if it does, replaces the input string with a Unix timestamp. Handling this case is debatable (*one could replace all forward slashes with some other character instead, to give uneducated users some idea as to what that template was when they used a forward slash in their infinite wisdom, not leave them stranded with a nebulous number*), though this way of handling it is probably the fastest, bar outright returning from the function without any notification to the user (that's another string to be translated anyhow).

This also fixes the problem of a forward slash causing a failure when trying to create a file under a directory that doesn't exist (try `foo/bar` as input string ...).

Edit: Pardon me, I added a commit to handle wrongslashes on some insane operating system I haven't used in quite a while.

### CODE REVIEW

**Comment by Qrox** [comment]
I think it'll be better just to disable the input of slashes (and other special characters, maybe in a platform-dependent way).

File: src/newcharacter.cpp [hunk 1]
```diff
@@ -2428,6 +2429,10 @@ void save_template( player *u )
     if( name.length() == 0 ) {
         return;
     }
+    if( name.find( '/' ) != std::string::npos || name.find( '\\' ) != std::string::npos ) {
```
Comment (BevapDin) [inline src/newcharacter.cpp]: You can use [`std::string::find_first_of`](http://en.cppreference.com/w/cpp/string/basic_string/find_first_of) to check for several characters in one function call.

File: src/newcharacter.cpp [hunk 2]
```diff
@@ -2428,6 +2429,10 @@ void save_template( player *u )
...
+        name.clear();
+        name.insert( 0, std::to_string( ( unsigned long long ) time( nullptr ) ) );
```
Comment (BevapDin) [inline src/newcharacter.cpp]: This is a very strange way to save `name = std::to_string( ... );`.

You could show a `popup` telling the use about the invalid characters and than just leave the function.

**Comment by Muffindrake** [comment]
@Qrox:

There is no trivial way to do that in a string input popup, given the facilities of the codebase, unless I missed something in a colossal manner.

@BevapDin:

There are a billion different ways to do the same thing in C++, I just can't be bothered to listen to whatever Bjarne thinks is cool to preach right now.

Let me know if I there's something I missed inside the PP directives.
Otherwise, the popup text may be subject to debate if RTFM sounds too harsh.

**Comment by Qrox** [comment]
I haven't tried myself, but the keybinding menu seems to be using `string_input_popup::ch_code_blacklist` to filter out `+`, `-` and `=` characteres, so I guess other characters can be filtered as well. Combining this with `string_input_popup::query_string(false)` and `string_input_popup::description()`, we can prompt the player immediately when they entered a invalid character, which is more user-friendly to my mind.

File: src/newcharacter.cpp [hunk 3]
```diff
@@ -31,6 +31,7 @@
 #include <vector>
 #include <algorithm>
 #include <cassert>
+#include <ctime>
```
Comment (Qrox) [inline src/newcharacter.cpp]: And this is no longer needed.

**Comment by Muffindrake** [comment]
@Qrox:
Indeed, there is a way. I opted for setting the description to `pls no special characters` and using `string_input_popup::query_string(true)` rather than false, as the blacklist filters invalid characters anyway, thus they aren't echoed to the terminal and the user gets what he sees.
I've added success/failure popups instead, giving feedback as to what has been done.

File: src/newcharacter.cpp [hunk 4]
```diff
@@ -2420,16 +2420,40 @@ trait_id Character::random_bad_trait()
...
+    spop.context( ctxt );
+    spop.ch_code_blacklist = fname_char_blacklist;
+
+    std::string name = spop.query_string( true );
+    if( name.empty() ) {
+        popup( _( "No filename given, template not saved" ) );
         return;
     }
+
     std::string playerfile = FILENAMES["templatedir"] + utf8_to_native( name ) + ".template";
```
Comment (Qrox) [inline src/newcharacter.cpp]: Might be less important, but I've heard that charset conversion functions like `utf8_to_native` here sometimes will introduce special characters into the converted string again. E.g. in Japanese Windows systems backslashes `\` shows up as the yen sign `¥`, and the unicode yen sign `¥(U+00A5)` may be converted to `\` in the conversion process. So if you want to be completely safe, it'll be nice to check for sanity for the return value of `utf8_to_native` too.

**Comment by Muffindrake** [comment]
I believe we've covered most bases here. I don't expect this PR to change any more, unless something spectacularly breaks with it.

**Comment by Leland** [comment]
@Muffindrake conflicts in newcharacter.cpp

File: src/newcharacter.cpp [hunk 5]
```diff
@@ -2420,15 +2420,48 @@ trait_id Character::random_bad_trait()
...
+
+    std::string name = spop.query_string( true );
+    if( name.empty() ) {
         return;
     }
-    std::string playerfile = FILENAMES["templatedir"] + utf8_to_native( name ) + ".template";
+
+    std::string native = utf8_to_native( name );
+#if (defined _WIN32 || defined __WIN32__)
+    if( name.find_first_of( "\"*/:<>?\\|"
```
Comment (BevapDin) [inline src/newcharacter.cpp]: Why is this checking `name`, but the error message says "Conversion ... to your native character set". It looks like it should check `native` instead.

**Comment by kevingranade** [comment]
Had to fix things up a little in the merge, which is my fault for leaving this so long, sorry about that.

**Pull request merged**